### PR TITLE
[feat] MADGRAD optimizer implementation

### PR DESCRIPTION
**Pull request by adefazio**
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
An implementation of the newly developed MADGRAD optimizer (https://arxiv.org/abs/2101.11075), a general purpose deep learning optimizer. It gives state-of-the-art results across a variety of computer vision and NLP problems.

### CODE REVIEW

File: fairscale/optim/madgrad.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+class MADGRAD(torch.optim.Optimizer):
+    """
+    MADGRAD_: A Momentumized, Adaptive, Dual Averaged Gradient Method for Stochastic 
+    Optimization.
+
+    .. _MADGRAD: https://arxiv.org/abs/2101.11075
+
+    MADGRAD is a general purpose optimizer that can be used in place of SGD or
+    Adam may converge faster and generalize better. Currently GPU-only.
```
Comment (blefaudeux) [inline fairscale/optim/madgrad.py:26]: "used in place of SGD or *and* Adam may converge" ?

File: fairscale/optim/madgrad.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/optim/madgrad.py:3]: I am naive on the license part. What's problem with MIT?

File: fairscale/optim/__init__.py [hunk 3]
```diff
@@ -19,3 +19,7 @@
     from .grad_scaler import GradScaler
 except ImportError:
     logging.warning("Torch AMP is not available on this platform")
+try:
+    from .madgrad import MADGRAD
+except ImportError:  # pragma: no cover
```
Comment (min-xu-ai) [inline fairscale/optim/__init__.py:24]: is this catch needed? for adascale we just directly import it without the catch.

File: stubs/torch/optim/__init__.pyi [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 from .sgd import SGD as SGD
 from .adam import Adam as Adam
+from .madgrad import MADGRAD as MADGRAD
```
Comment (min-xu-ai) [inline stubs/torch/optim/__init__.pyi:5]: is this needed? we only need it for torch's optimizer here as a stub

**Review by min-xu-ai — COMMENTED** [review]
Thanks for including the doc and tests. Several high level comments.

**Pull request closed without merging**